clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -306,6 +306,36 @@ static int64_t getIntExprValue(const Expr *E, ASTContext &Ctx) {
   return E->getIntegerConstantExpr(Ctx)->getSExtValue();
 }
 
+// Collect function attributes related to SYCL
+static void collectSYCLAttributes(Sema &S, FunctionDecl *FD,
+                                  llvm::SmallVector<Attr *, 4> &Attrs,
+                                  bool DirectlyCalled = true) {
+  if (!FD->hasAttrs())
+    return;
+
+  llvm::copy_if(FD->getAttrs(), std::back_inserter(Attrs), [](Attr *A) {
+    return isa<IntelReqdSubGroupSizeAttr, ReqdWorkGroupSizeAttr,
```

Comment:
Is it expected that this list will grow more in the future? If so, this seems like it's going to be error-prone as people add new SYCL attributes because it'll be easy to forget to update this list.

\<thinking out loud>One idea would be to introduce a new attribute base class in Attr.td, similar to `TargetSpecificAttr`, that acts as a mix-in base class for SYCL attributes (it would have to be a base class of the actual semantic attribute generated by tablegen as well so that it works with the LLVM casting machinery). The attributes in there would then inherit from `SYCLAttr` as well as whatever other base class makes sense (`StmtAttr`, `InheritableAttr`, etc). Whenever we need to do special processing for a large group of SYCL attributes like here, we use the `SYCLAttr` base class to query the information. e.g., `if (const auto *SA = dyn_cast<SYCLAttr>(A)) { return SA->isCopiedToKernel(); }`</thinking out loud>

FWIW, I think the code can be committed as-is, but we may want to consider adding a FIXME comment to it and opening an issue.

Reply:
OK, I have added a FIXME comment and will create an issue after this code is merged.